File: backoffice/src/shared/infrastructure/events/DomainEventsAggregator.py
"""
 *
 * Libraries 
 *
"""

from .DomainEventsInformation             import DomainEventsInformation
from .DomainEventInformation              import DomainEventInformation
from .DomainEventSubscribersLocator       import DomainEventSubscribersLocator
from .DomainEventsLocator                 import DomainEventsLocator
from injector                             import Module
from ..InjectorDependecyServiceAggregator import InjectorDependecyServiceAggregator

"""
 *
 * Class
 *
"""

class DomainEventsAggregator( Module ):

    """
     *
     * Methods 
     *
    """

    def configure( self, binder ) -> None:
        # Variables
        eventsInformation      : list[DomainEventInformation]
        eventSubscriberClasses : list[type]
        eventClasses           : list[type]
        # Code        
        eventsInformation = []
        eventClasses      = DomainEventsLocator.locate()
        for eventClass in eventClasses:
            eventSubscriberClasses = DomainEventSubscribersLocator.locate( eventClass )
            eventsInformation.append( DomainEventInformation( 
                eventClass, 
                eventSubscriberClasses 
            ) )
        InjectorDependecyServiceAggregator.addSingleton( 
            binder, 
            DomainEventsInformation, 
            DomainEventsInformation( eventsInformation ) 
        )

File: preparation/src/shared/infrastructure/events/DomainEventsAggregator.cs
/* 
 * 
 * Class
 *
*/

namespace preparation.src.shared.infrastructure {

    public static class DomainEventsAggregator {

        /* 
         * 
         * Methods
         *
        */

        public static IServiceCollection addDomainEvents( this IServiceCollection services ) {
            // Variables
            List<DomainEventInformation> eventsInformation;
            List<Type>                   eventSubscriberClasses;
            List<Type>                   eventClasses;
            // Code
            eventsInformation = new List<DomainEventInformation>();
            eventClasses      = DomainEventsLocator.locate();
            foreach( Type eventClass in eventClasses ) {                
                eventSubscriberClasses = DomainEventSubscribersLocator.locate( eventClass );
                eventsInformation.Add( new DomainEventInformation( 
                    eventClass, 
                    eventSubscriberClasses 
                ) );                
                foreach( Type eventSubscriberClass in eventSubscriberClasses ) {
                    services.AddTransient( eventSubscriberClass, eventSubscriberClass );
                }
            }
            services.AddTransient( s => new DomainEventsInformation( eventsInformation ) );
            return services;
        }

    }
}
